rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+   functional index is registered in a separate index metadata file, to keep track of the relationship between the data
+   column and the functional Index. 
+3. Each functional index will be a partition inside the Hudi metadata table (MT).
+4. No data is materialized to the data files for the functional index, i.e., the data files do not contain a new data
+   column corresponding to the Functional Index, to save storage space.
+5. When query engine makes a logical plan for a query, Hudi intercepts the predicate that relates to a functional index,
+   looks for the corresponding index in the MT, and applies data skipping based on one or more indexes.
+6. In order to be engine-agnostic, the design should not make assumptions about any particular engine.
+
+### Components
+
+We discuss the design and implementation details of each component to realize the Functional Index.
+
+#### SQL
+
+A new keyword `FUNCTION` is introduced to indicate that a functional index is intended to be created. This keyword is
+ANSI-compliant and also present in [spark-sql](https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html).
+
+```sql
+-- PROPOSED SYNTAX WITH FUNCTION KEYWORD --
+CREATE
+[FUNCTION] INDEX index_name ON table_name [ USING index_type ] ( { column_name | expression } );
+-- Examples --
+CREATE FUNCTION INDEX last_name_idx ON employees (UPPER(last_name)); -- functional index using column stats for UPPER(last_name)
+CREATE FUNCTION INDEX datestr ON hudi_table (DATE_FORMAT(timestamp, '%Y-%m-%d')); -- functional index using column stats for DATE_FORMAT(timestamp, '%Y-%m-%d')
+CREATE INDEX city_id_idx ON hudi_table (city_id); -- usual column stats within MT column_stats partition
+CREATE FUNCTION INDEX hour_of_day ON hudi_table USING BITMAP (DATE_FORMAT(timestamp, '%H')); -- functional index using bitmap for DATE_FORMAT(timestamp, '%H')
+CREATE FUNCTION INDEX income_idx ON employees (salary + (salary*commission_pct)); --  functional index using column stats for given expression
+-- NO CHANGE IN DROP INDEX --
+DROP INDEX last_name_idx;
+```
+
+`index_name` - Required, should be validated by parser. The name will be used for the partition name in MT.
+
+`index_type` - Optional, column\_stats if none provided and there are no functions and expressions in the command. Valid
+options could be BITMAP, COLUMN\_STATS, LUCENE, etc. If index\_type is not provided, and there are functions or
+expressions in the command then a functional index using column stats will be created.
+
+`expression` - simple scalar expressions or sql functions.
+
+#### Functional Index Metadata
+
+For each functional index, store a separate metadata with index details. This should be efficiently loaded. One option
+is to store the below metadata in `hoodie.properties`. This way all index metadata can be loaded with the table config.
+But, it would be better to not overload the table config. Let `hoodie.properties` still hold the list of indexes (MT
```

Comment:
Can we store a property in `hoodie.properties` to the path of the index metadata, for extensibility and not assuming its `.hudi_index`. also just `.index_defs` or something like that that makes it clear this is not the actual index, but just the definitions.

Reply:
`hoodie.properties` will have the MT partition name corresponding to each index in `hoodie.table.metadata.partitions` property (this exists even today). We can store the full path to the index in `.index_defs`. Do you see some benefit of storing the path in `hoodie.properties`?

Reply:
Updated rfc with a new table config in hoodie.properties - `hoodie.table.index_defs.path`

Reply:
changed to `hoodie.table.index.defs.path` ?